### PR TITLE
module-level factory functions

### PR DESCRIPTION
**Pull request by rgabbard**
Partially addresses #12

### CODE REVIEW

**Comment by rgabbard** [comment]
This is a start on #12 . A few notes:

* to test the error messages on different versions of Python, we probably need to switch to testing with `tox`. I am inclined to push this testing (and therefore testing the order checks for dicts) to a different issues.
* I think we should factory implementing `check_isinstance` to its own issue because there are some tricky subtleties there (e.g. if the input is already an immutable set but your requested type is a sub-type of the type recoreded on it, what is the right thing to do?) and no one currently uses it.

**Comment by rgabbard** [comment]
@nicomitchell : can you add tests to this for the order check case of `AbstractSet`?

File: immutablecollections/_immutableset.py [hunk 1]
```diff
@@ -28,6 +29,93 @@
...
+            if isinstance(iterable, KeysView) or isinstance(iterable, ValuesView) or isinstance(
+                    iterable, ItemsView):
+                    raise ValueError("Attempting to initialize an ImmutableSet from "
+                                     "a dict view. On this Python version, this probably loses "
+                                     "determinism in iteration order.  If you don't care "
+                                     "or are otherwise sure your input has determinstic "
+                                     "iteration order, specify disable_order_check=True")
+
+    # iterable might not be safe to iterate multiple times, so we copy it
+    values = tuple(iterable)
```
Comment (ConstantineLignos) [inline immutablecollections/_immutableset.py]: I think we don't need the tuple conversion, since we only iterate once.

**Comment by ConstantineLignos** [comment]
Another general comment on something that isn't in the diff so I can't leave a review: can we change `iteration_order` inside `_FrozenSetBackedImmutableSet` to just be `Sequence[T]` and not do any conversion?

Since it's effectively a private constructor, we can trust that the factory method is never sharing the iteration order with the caller, even if it is mutable.

Current implementation is:
```
    _iteration_order: immutablelist.ImmutableList[T] = attrib(
        converter=immutablelist.ImmutableList.of, cmp=False, hash=False  # type: ignore
    )
```

File: immutablecollections/_immutableset.py (hunk 1 above)
Comment (rgabbard) [reply immutablecollections/_immutableset.py]: Yep, this a relic of the type checking option, where we do need to iterate twice. But when/if the type checking is implemented, we can just do type checking on the constructed object and avoid the tuple-ization.

**Comment by rgabbard** [comment]
@ConstantineLignos : actually, is there any reason to have `as_list()` now that we just implement `Sequence`? I think it was added before that was the case

**Comment by rgabbard** [comment]
@ConstantineLignos : I think an event nicer option is for `ImmutableSet` to just sub-class tuple and skip having a separate `iteration_order` field (instead passing the order to `super`).  This may cause a slight slowdown on creation, but should save time on access and iteration.

**Comment by rgabbard** [comment]
(but this should be checked using benchmarks)

**Comment by ConstantineLignos** [comment]
I'm fine with killing off `as_list`. If subclassing tuple makes sense, go for it. I'd like to explore the idea of a dict-backed implementation (see https://github.com/isi-vista/immutablecollections/issues/14#issuecomment-457265569) at the same time as when exploring subclassing tuple. The dict-backed one will should have the advantage of having the backing dict handle both containment and order in the dict itself, so the factory method would be different for maximum efficiency.

(Just in case anyone is wondering, the Python 3.6 dict implementation keeps the order of *first* insertion as you might expect, just like `ImmutableSet`, see below.)

```
>>> list(dict({item: None for item in [4, 3, 2, 1, 4]}))
[4, 3, 2, 1]
```

**Comment by rgabbard** [comment]
@ConstantineLignos : I experimented a little with using a `dict` in the factory method:

```
unique_values = tuple({ val : None for val in iterable}.keys())
return _FrozenSetBackedImmutableSet(
                frozenset(unique_values), unique_values)
...


  In increases the time to initialize from a 3-element collection by ~5% but doubles performance on a 10k-element collection, so I think it is a win.  

One challenge with using the `dict` to actually back the set is that you still need the `tuple` for iteration order because the `Dict`'s `KeyView` just extends `Collection`, not `Sequence`.  So the tradeoff becomes the time to initialize the `frozenset` for containment checks vs just keeping the `dict` vs. the (probable) increased memory usage of `dict`

**Comment by ConstantineLignos** [comment]
@rgabbard Is there anything we'd need to do before merging if we stick with the frozenset backing right now?

**Comment by rgabbard** [comment]
@ConstantineLignos : I have some local changes on this branch to sync up with this before merging. Poke me if I haven't taken care of it by tomorrow morning

**Comment by ConstantineLignos** [comment]
Sounds good. I'm on vacation at the start of next week so I'd like to get this and https://github.com/isi-vista/vistautils/issues/44 sorted this week before I go.

File: immutablecollections/_immutableset.py [hunk 2]
```diff
@@ -6,29 +6,112 @@
...
+            ):
+                raise ValueError(
+                    "Attempting to initialize an ImmutableSet from "
+                    "a dict view. On this Python version, this probably loses "
+                    "determinism in iteration order.  If you don't care "
+                    "or are otherwise sure your input has determinstic "
+                    "iteration order, specify disable_order_check=True"
+                )
+
+    # iterable might not be safe to iterate multiple times, so we copy it
```
Comment (ConstantineLignos) [inline immutablecollections/_immutableset.py]: Don't need this comment anymore.

**Comment by rgabbard** [comment]
I'm all in favor of ditching the type checking (which I think also means, on a future PR, ditching `ImmutableList`, which no longer offers any advantage over `tuple`).  I'll drop the TODO comments related to it.

**Comment by rgabbard** [comment]
I'm ignoring `codecov`'s comments about test coverage for now because it is largely due to me switching tests to use the new factory methods, leaving parts of the builder and `.of` methods untested.  This code is slated for deprecation shortly, so I'm not worried about it.

**Pull request merged**